oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/resourceserver/authentication/JwtAccessTokenAuthenticationProvider.java

```diff
@@ -0,0 +1,88 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.resourceserver.authentication;
+
+import org.springframework.security.authentication.AuthenticationProvider;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.OAuth2Error;
+import org.springframework.security.oauth2.core.OAuth2ErrorCodes;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.security.oauth2.jwt.JwtDecoder;
+import org.springframework.security.oauth2.jwt.JwtException;
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken;
+import org.springframework.util.Assert;
+
+/**
+ * An {@link AuthenticationProvider} implementation of the OAuth2 Resource Server Bearer Token when using Jwt-encoding
+ * <p>
+ * <p>
+ * This {@link AuthenticationProvider} is responsible for decoding and verifying a Jwt-encoded access token,
+ * returning a Jwt claims set as part of the {@see Authentication} statement.
+ *
+ * @author Josh Cummings
+ * @author Joe Grandja
+ * @since 5.1
+ * @see JwtAccessTokenVerifier
+ * @see AuthenticationProvider
+ * @see JwtDecoder
+ */
+public class JwtAccessTokenAuthenticationProvider implements AuthenticationProvider {
+	private final JwtDecoder jwtDecoder;
+	private final JwtAccessTokenVerifier jwtVerifier;
+
+	public JwtAccessTokenAuthenticationProvider(JwtDecoder jwtDecoder) {
+		this(jwtDecoder, new JwtAccessTokenVerifier());
+	}
+
+	public JwtAccessTokenAuthenticationProvider(JwtDecoder jwtDecoder,
+												JwtAccessTokenVerifier jwtVerifier) {
+		Assert.notNull(jwtDecoder, "jwtDecoder is required");
+		Assert.notNull(jwtVerifier, "jwtVerifier is required");
+
+		this.jwtDecoder = jwtDecoder;
+		this.jwtVerifier = jwtVerifier;
+	}
+
+	@Override
+	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+		PreAuthenticatedAuthenticationToken preAuthentication =
+				(PreAuthenticatedAuthenticationToken) authentication;
+
+		Jwt jwt;
+		try {
+			jwt = this.jwtDecoder.decode(String.valueOf(preAuthentication.getPrincipal()));
+		} catch (JwtException failed) {
+			OAuth2Error invalidRequest =
+					new OAuth2Error(
+							OAuth2ErrorCodes.INVALID_REQUEST,
+							failed.getMessage(),
+							"https://tools.ietf.org/html/rfc6750#section-3.1");
+
+			throw new OAuth2AuthenticationException(invalidRequest, failed);
+		}
+
+		this.jwtVerifier.verify(jwt.getClaims());
+
+		return new JwtAccessTokenAuthenticationToken(jwt);
```

Comment:
I'm not convinced that we need to make use of `GrantedAuthority` on the Resource Server side. `scope` is essentially the only clearly documented authority in the spec. Although there may be authority-like attributes in the token that could be mapped to one or more `GrantedAuthority` instances, this would provide 2 options for the user to configure the authorization rules:

1) `hasAuthority('message-read')`

OR

2) `hasAttribute('authority', 'message-write')`

It may be confusing to the user on which option to use?

I think we should definitely have a discussion on this and figure out the path to take.

Reply:
I'll summarize here an off-line conversation that @jgrandja and I had about this point.

The concern boils down to what are the correct mechanisms for ensuring that a class like this:

```java
public interface MySecuredRepository {
    Message getMessage(String id);
}
```

Doesn't have a direct reference to `@oauth2` like this:

```java
public interface MySecuredRepository {
    @PreAuthorize("hasScope('message.read')")
    Message getMessage(String id);
}
```

In order to accomplish this, we will either need to provide copy the scope attribute preemptively into a set of granted authorities or we will need to provide a hook and possibly a default implementation to allow devs to do this themselves (thus, an authorities mapper).

For the time being, this PR will stand, though I will log an issue to bring this support back in. It may rematerialize as an authorities mapper or maybe something else, and that issue will allow us to flesh it out.